---
title: "COLOMBOS DB E. Coli expression exploratory data analysis"
author: "Emanuel Becerra Soto"
date: "June 05, 2019"
output: pdf_document
---


```{r}
suppressMessages(library("tidyverse"))
library(ggthemes)
library(viridis)
library(ggrepel)
library(NbClust)
library(forcats)
library(scales)

# For reproducibility
set.seed(243014)

quantile_plot <- function(data, feature){
  library(tidyverse)
  library(ggrepel)
  n <- nrow(data)
  percentiles <- (1:n - 0.5) / n
  values <- quantile( x = data[[ feature ]] , probs = percentiles)
  quantiles <- quantile(data[[ feature ]])
  IQ1 <- quantiles[2]
  Median <- quantiles[3]
  Q3 <- quantiles[4]
  normal_values <- qnorm(seq(0.01,1,0.01), mean = mean(data[[ feature ]]), sd = sd(data[[ feature ]]))
  unif_values <- qunif(seq(0.01,1,0.01), max = max(data[[ feature ]]), min = min(data[[ feature ]]))
  percentiles2 <- c(percentiles, 0.25, 0.50, 0.75)
  values2 <- c(values, IQ1, Median, Q3)
  to_plot <- tibble(percentiles2, values2)
  names(to_plot) <- c('percentiles', 'values')
  theoretical <- tibble(perc =  seq(0.01,1,0.01) , normal_values, unif_values)
  last_three <- (nrow(to_plot) - 2) : nrow(to_plot)
  theoretical_tidy <- theoretical %>%
    gather(normal_values, unif_values, key = 'distribution', value = 'values' )
  ggplot(to_plot, aes(x = percentiles, y = values))+
    geom_point(size  = 2)+
    geom_line(data = theoretical_tidy,
              aes(x = perc, y=values, color = distribution),
              size  = 1.5,
              alpha = 0.6)+
    geom_point(data = to_plot[ last_three, ],
               fill = 'red', color = 'red', shape = 23, size = 2.5 )+
    geom_text_repel( data = to_plot[ last_three, ],
                     aes(label = c('Q1','Median','Q3')),
                     vjust        = -5,
                     segment.size = 0.2)+
    ggtitle(paste(feature,'Quantile Plot'))+
    scale_color_brewer(type = 'qual', palette = 6, name = 'Distribution', labels=c('Normal', 'Uniform'))
}
```

## Introduction

Exploratory thesis

What is COLOMBOS DB?
Integrates transcriptomics data
for several prokaryotic model-organisms

Micro-array and RNA-Seq

The motivation of the project is to address the challenge of novel
biological discoveries derived from the future collection of ChIP-seq
and RNA-seq experiments for every single transcription factor in E. coli,
using machine learning approaches.
More precisely we will focus on searching for correlation
of labels of co-expressed/co-regulated genes using the COLOMBOS collection as a
first step. This will prepare us to use the future ChIP-seq and RNA-seq
experiments to propose new biological knowledge in E. coli.


## Reading the data


```{r}
expr_file <- 'ecoli_expr.tsv'

# Reading the data
ecoli_expr <- read_tsv(expr_file)
dim(ecoli_expr)
head(ecoli_expr)

# Conditions
n_cond <- 4077
n_gene <- nrow(ecoli_expr)
```


```{r}
# Counting NA's per column
na_count_per_condition <- sapply(ecoli_expr[4:4080], function(col) {
  sum(is.na(col))
})

miss_cond <- tibble(cond = 1:n_cond,
                    abs_miss = na_count_per_condition,
                    rel_miss = na_count_per_condition/n_cond)

# Plotting missing conditions per gene
ggplot(miss_cond, aes(x = cond, y = rel_miss*100))+
  geom_step(color = viridis_pal()(1))+
  scale_y_continuous(
    sec.axis = sec_axis(~ . * nrow(miss_cond)/100, name = "Absolute Missing"))+
  labs(title="Missing Values % per Condition", 
       subtitle="Max up to 21.8% missing",
       x = "Conditions",
       y = "% Missing",
       caption="source: COLOMBOS DB")
```

## Removing missing values

```{r}
# Removing missing values
# Substituting by the median value of the condition
NA2median <- function(x) replace(x, is.na(x), median(x, na.rm = TRUE))
ecoli_expr <- replace(ecoli_expr, TRUE, lapply(ecoli_expr, NA2median))
```

## Summary of conditions

```{r}
# Summary of first 20 columns
# summary(ecoli_expr[1:20])
```


```{r}
numeric_ecoli_expr <- ecoli_expr[4:ncol(ecoli_expr)]

max_e <- sapply(numeric_ecoli_expr, max)
min_e <- sapply(numeric_ecoli_expr, min)
med_e <- sapply(numeric_ecoli_expr, median)
sd_e <- sapply(numeric_ecoli_expr, sd)

# Summary per condition
ecoli_expr_summary <- tibble( condition = 1:length(med_e), med_fold = med_e,
                          sd_fold = sd_e, max_fold = max_e, min_fold = min_e )
```


```{r}
ggplot(ecoli_expr_summary)+
  geom_point( aes(med_fold, sd_fold), alpha = 1/10 )+
  xlab('Condition Median')+
  ylab('condition SD')+
  ggtitle('Conditions Fold Change: Median vs SD')
```


```{r}
ggplot(ecoli_expr_summary)+
  geom_point( aes(max_fold, sd_fold), alpha = 1/10 )+
  xlab('Condition Max')+
  ylab('condition SD')+
  ggtitle('Conditions Fold Change: Max vs SD')
```


```{r}
ggplot(ecoli_expr_summary)+
  geom_point( aes(min_fold, sd_fold), alpha = 1/10 )+
  xlab('Condition Min')+
  ylab('condition SD')+
  ggtitle('Conditions Fold Change: Min vs SD')
```

```{r}
ggplot(ecoli_expr_summary)+
  geom_point( aes(min_fold, max_fold), alpha = 1/10 )+
  xlab('Condition Min')+
  ylab('condition Max')+
  ggtitle('Conditions Fold Change: Min vs Max')
```

## Example Conditions

Number 1
E-TABM-103.14.ch1-vs-E-TABM-103.3.ch1

MEDIUM.LB:1
TIME:186min
GROWTH.EXPONENTIAL:1
STRAIN.RP437:1

```{r}
ggplot(ecoli_expr, aes(x = 1:nrow(ecoli_expr), y = `1`))+
  geom_point(size = 0.5, color = '#9e3d40')+
  geom_text_repel( aes(label = ifelse(`1` >= 1.8, `Gene name`, '')) )+
  geom_text_repel( aes(label = ifelse(`1` <= -1.8, `Gene name`, '')) )+
  labs(title="E-TABM-103.14.ch1-vs-E-TABM-103.3.ch1", 
     subtitle="The marked genes were above or below +/- 1.8 FC",
     x = "Gene",
     y = "Fold Change Condition 1",
     caption="source: COLOMBOS DB")
```

```{r}
# Quantile plot
quantile_plot(ecoli_expr, '1')
```


Number 2
E-TABM-103.19.ch1-vs-E-TABM-103.3.ch1

MEDIUM.LB:1
TIME:186min
GROWTH.EXPONENTIAL:1
STRAIN.RP437:1

```{r}
ggplot(ecoli_expr, aes(x = 1:nrow(ecoli_expr), y = `2`))+
  geom_point(size = 0.5, color = '#9e3d40')+
  geom_text_repel( aes(label = ifelse(`2` >= 1.8, `Gene name`, '')) )+
  geom_text_repel( aes(label = ifelse(`2` <= -1.8, `Gene name`, '')) )+
  labs(title="E-TABM-103.19.ch1-vs-E-TABM-103.3.ch1", 
   subtitle="The marked genes were above or below +/- 1.8 FC",
   x = "Gene",
   y = "Fold Change Condition 2",
   caption="source: COLOMBOS DB")
```

```{r}
# Quantile plot
quantile_plot(ecoli_expr, '2')
```

## Quartiles of fold change per gene

```{r}
# Gene quartiles
gene_quartiles <- apply(numeric_ecoli_expr, MARGIN = 1, quantile) %>%
  t() %>%
  as_tibble() %>%
  mutate(gene = ecoli_expr$`Gene name`, num_gene = 1:n_gene) %>%
  gather(`0%`, `25%`, `50%`, `75%`, `100%`,
       key = 'percentile', value = 'fold_change')

# Percentil as a factor
gene_quartiles$percentile <- factor(gene_quartiles$percentile)
# Reorder Levels to get a better plot
gene_quartiles$percentile <- factor(gene_quartiles$percentile,
       levels = levels(gene_quartiles$percentile)[c(2,1,5,3,4)])
head(gene_quartiles)
```


```{r}
# Geom Line
ggplot(gene_quartiles, aes(x = 1:n_gene))+
  geom_point(aes(x = num_gene, y = fold_change, color = percentile), alpha = 1/3, size = 0.7)+
  scale_color_viridis_d()+
  labs(title = 'Summary per Gene: Fold Change Quartiles',
      x = 'Gene',
      y = 'Fold Change')
```

```{r}
# Area Fill
ggplot(gene_quartiles, aes(x = 1:n_gene))+
  geom_area(aes(x = num_gene, y = fold_change, fill = percentile) )+
  scale_fill_viridis_d()+
  labs(title = 'Summary per Gene: Fold Change Quartiles',
      x = 'Gene',
      y = 'Fold Change')
```


```{r}
# Area Fill Zoom
ggplot(gene_quartiles, aes(x = 1:n_gene))+
  geom_area(aes(x = num_gene, y = fold_change, fill = percentile) )+
  coord_cartesian(ylim = c(-1,1))+
  scale_fill_viridis_d()+
  labs(title = 'Summary per Gene: Fold Change Quartiles',
      x = 'Gene',
      y = 'Fold Change')
```


```{r}
# What Geneid/Constrast_id are?
plot(ecoli_expr$`Geneid/Contrast_id`,
     xlab = 'Order',
     ylab = 'GeneID / Constrast',
     main = 'The column Geneid/Contrast_id has this pattern',
     col = 'gray60'
)
```

## Saving tables

```{r}
# Saving the table
# Inverting to take the conditions as observations
cond_gene <- as_tibble( t( numeric_ecoli_expr ) )
names(cond_gene) <- ecoli_expr$`Gene name`

head(cond_gene)

# Saving to disk
# saveRDS(cond_gene, file = 'conditions_vs_genes.RDS')

# Creating gene vs condition table
# gene_cond <- as_tibble(t(cond_gene))
# colnames(gene_cond) <- paste('C', 1:ncol(gene_cond), sep = '')
# gene_cond$gene <- names(cond_gene) 
# gene_cond <- gene_cond[ , c(ncol(gene_cond),1:ncol(gene_cond)-1) ]
# gene_cond

# Saving gene vs condition table
# saveRDS(gene_cond , 'genes_vs_conditions.RDS')

# Loading tables
# condition_gene <- readRDS('genes_vs_conditions.RDS')
gene_condition <- readRDS('genes_vs_conditions.RDS')

head(gene_condition)
```

## PCA and Clustering

```{r}
PCA_reduce_dimension <- function(data, perct_tot_var = 0.95){
  pca_re <- prcomp(data)
  # Loadings
  var_per_comp <- pca_re$sdev^2
  # Cutoff for dim reduction
  var_cumsum <- cumsum(var_per_comp / sum(var_per_comp))
  # plot(var_cumsum)
  cut_idx <- which(var_cumsum >= perct_tot_var)[1]
  dim_reduc <- pca_re$x[,1:cut_idx]
  print(paste('Your data was reduced from',
              ncol(data), 'columns', 'to', cut_idx,
              'capturing', perct_tot_var * 100, 'percent of the variance'))
  return(list(reduced = dim_reduc, pca = pca_re))
}

# Running PCA to reduce dimensions
pca_results <- PCA_reduce_dimension(gene_condition[2:ncol(gene_condition)],
                                    perct_tot_var = 0.80)

# Reduced table
reduced_gene_cond <- as_tibble(pca_results$reduced)
```

```{r}
var_pca <- pca_results$pca$sdev ^ 2
total_var_pca <- sum(var_pca)
cum_var_pca <- cumsum(var_pca / total_var_pca)

plot(cum_var_pca,
     xlab = 'Principal Components',
     ylab = 'Accumulated Variance',
     main = 'PCA Variance Gene vs Conditions',
     col = 'gray60',
     frame.plot = FALSE # Remove the frame
)
```


```{r}
# Variance captured by the first two components
(PCA1_explained <- round( var_pca[1] / total_var_pca * 100, 2 ))
(PCA2_explained <- round( var_pca[2] / total_var_pca * 100, 2 ))
```

Using NbClust to test for the number of gene clusters
thus getting the best values for k (number of clusters)
in k-means algorithm

```{r}
# Heavy computational line
# Time to run aprox 1 hour
# nc_ALL <- NbClust(reduced_gene_cond, min.nc=2, max.nc=30, method="kmeans", index = 'all')
# Saving the results
# saveRDS(nc_ALL, 'number_clusters_ALL_gene_cond.RDS')

# Loading the results
nc_ALL <- readRDS('number_clusters_ALL_gene_cond.RDS')
```

```{r}
nc_gene <- as_tibble(t(nc_ALL$Best.nc))
nc_gene$Method <- colnames(nc_ALL$Best.nc)
nc_gene <- select(nc_gene, Method, everything()) %>%
  arrange(Number_clusters)
nc_gene[1:3, 2] <- 1
nc_gene
```

```{r}
ggplot(nc_gene)+
  geom_bar(aes(x = fct_infreq(as.character(Number_clusters))),
           size = 0.4,
           fill = '#F79C91')+
  scale_y_continuous(breaks= pretty_breaks())+
  labs(title="Number of gene clusters according to 26 Criteria",
     x = "Number of Clusters",
     y = "Counts")+
  theme_hc()
```

### K-means

```{r}
# K means with 2, 3 and 26 clusters
clust2 <- kmeans(reduced_gene_cond, centers = 2, nstart = 50, iter.max = 30)
clust3 <- kmeans(reduced_gene_cond, centers = 3, nstart = 50, iter.max = 30)
clust26 <- kmeans(reduced_gene_cond, centers = 26, nstart = 50, iter.max = 30)
```

```{r}
gene_cond_clusters <- mutate(reduced_gene_cond,
       gene = ecoli_expr$`Gene name`,
       group2 = clust2$cluster,
       group3 = clust3$cluster,
       group26 = clust26$cluster) %>%
  select(gene, group2, group3, group26, everything())
```

### Plots PCA and Clustering

```{r}
ggplot(gene_cond_clusters)+
  geom_point(aes(x = PC1, y = PC2, color = as.character(group2)),
             size = 0.7)+
  xlab(paste('PCA1 ', '%', PCA1_explained, sep = ''))+
  ylab(paste('PCA2 ', '%', PCA2_explained, sep = ''))+
  scale_color_brewer(type = 'qual')+
  labs(title = 'PCA of E. Coli genes over ~4000 conditions',
       color = 'Cluster')+
  geom_point(data = as_tibble(clust2$centers), aes(x = PC1, y = PC2),
             color = 'red', size = 3, shape = 18)
```


```{r}
ggplot(gene_cond_clusters)+
  geom_point(aes(x = PC1, y = PC2, color = as.character(group3)),
             size = 0.7)+
  xlab(paste('PCA1 ', '%', PCA1_explained, sep = ''))+
  ylab(paste('PCA2 ', '%', PCA2_explained, sep = ''))+
  scale_color_brewer(type = 'qual')+
  labs(title = 'PCA of E. Coli genes over ~4000 conditions',
       color = 'Cluster')+
  geom_point(data = as_tibble(clust3$centers), aes(x = PC1, y = PC2),
             color = 'red', size = 3, shape = 18)
```

```{r}
custom26 <- c("#FF7F00", "#FC8D59", "#A65628", "#999999",
              "#BEAED4", "#66A61E", "#BF5B17", "#FF4DCC",
              "#4DAF4A", "#7570B3", "#666666", "#FDC086",
              "#FFFF33", "#F0027F", "#000000", "#F781BF",
              "#38FF26", "#E7298A", "#E41A1C", "#D95F02",
              "#99D594", "#FFFFBF", "#377EB8", "#984EA3",
              "#386CB0", "#A6761D")

ggplot(gene_cond_clusters)+
  geom_point(aes(x = PC1, y = PC2, color = as.character(group26)),
             size = 0.7)+
  xlab(paste('PCA1 ', '%', PCA1_explained, sep = ''))+
  ylab(paste('PCA2 ', '%', PCA2_explained, sep = ''))+
  scale_color_manual(values = custom26)+
  labs(title = 'PCA of E. Coli genes over ~4000 conditions',
       color = 'Cluster')
```


```{r}
ggplot(gene_cond_clusters)+
  geom_point(aes(x = PC1, y = PC2, color = as.character(group26)),
             size = 0.8)+
  xlab(paste('PCA1 ', '%', PCA1_explained, sep = ''))+
  ylab(paste('PCA2 ', '%', PCA2_explained, sep = ''))+
  scale_color_manual(values = custom26)+
  labs(title = 'PCA of E. Coli genes over ~4000 conditions',
       color = 'Cluster')+
  coord_cartesian(xlim = c(-20,20), ylim = c(-20,20))
```


```{r}
ggplot(gene_cond_clusters)+
  geom_bin2d(aes(x = PC1, y = PC2),
             bins = 100)+
  xlab(paste('PCA1 ', '%', PCA1_explained, sep = ''))+
  ylab(paste('PCA2 ', '%', PCA2_explained, sep = ''))+
  scale_fill_viridis_c()+
  labs(title = 'PCA of E. Coli genes over ~4000 conditions')+
  coord_equal(ratio=1)
```


```{r}
explore_pcs3 <- function(pcx, pcy){
  ggplot(gene_cond_clusters)+
  geom_point(aes_string(x = pcx, y = pcy, color = 'group3'),
             size = 0.7)+
  scale_color_viridis_c()+
  labs(title = 'PCA of E. Coli genes over ~4000 conditions',
       color = 'Cluster')
  }
```


```{r}
explore_pcs3('PC1', 'PC3')
explore_pcs3('PC1', 'PC4')
explore_pcs3('PC2', 'PC3')
explore_pcs3('PC2', 'PC4')
```

